compiler/rustc_interface/src/queries.rs

```diff
@@ -107,3 +108,32 @@ impl Linker {
         codegen_backend.link(sess, codegen_results, self.metadata, &self.output_filenames)
     }
 }
+
+/// Checks if the given `DefId` refers to an item that is unnamable, such as one defined in a const block.
+///
+/// # Example
+///
+/// ```
+/// pub fn f() {
+///     // In here, `X` is not reachable outside of `f`, making it unnamable.
+///     pub struct X;
+/// }
+/// ```
+pub(crate) fn is_unnamable(tcx: TyCtxt<'_>, did: DefId) -> bool {
+    let mut cur_did = did;
+    while let Some(parent) = tcx.opt_parent(cur_did) {
+        match tcx.def_kind(parent) {
+            // items defined in these can be linked to, as long as they are visible
+            DefKind::Mod | DefKind::ForeignMod => cur_did = parent,
+            // items in impls can be linked to,
+            // as long as we can link to the item the impl is on.
+            // since associated traits are not yet a thing,
+            // it should not be possible to refer to an impl item if
+            // the base type is not namable.
+            DefKind::Impl { .. } => return false,
+            // everything else does not have docs generated for it
+            _ => return true,
+        }
+    }
+    return false;
+}
```

Comment:
Could you put this in `compiler/rustc_interface/src/passes.rs`? `compiler/rustc_interface/src/queries.rs` refers to a completely different kind of queries, that we should eventually extinguish.